src/vs/workbench/parts/explorers/browser/treeExplorerActions.contribution.ts

```diff
@@ -0,0 +1,57 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+'use strict';
+
+import * as nls from 'vs/nls';
+import { TPromise } from 'vs/base/common/winjs.base';
+import { Registry } from 'vs/platform/platform';
+import { IWorkbenchActionRegistry, Extensions as ActionExtensions } from 'vs/workbench/common/actionRegistry';
+import { SyncActionDescriptor } from 'vs/platform/actions/common/actions';
+import { Action } from 'vs/base/common/actions';
+import { IQuickOpenService, IPickOpenEntry } from 'vs/workbench/services/quickopen/common/quickOpenService';
+import { IActivityService } from 'vs/workbench/services/activity/common/activityService';
+
+const registry = Registry.as<IWorkbenchActionRegistry>(ActionExtensions.WorkbenchActions);
+
+export class ToggleExternalViewletAction extends Action {
+	public static ID = 'workbench.action.customTreeExplorer.toggle';
+	public static LABEL = nls.localize('toggleCustomExplorer', 'Toggle Custom Explorer');
+
+	constructor(
+		id: string,
+		label: string,
+		@IQuickOpenService private quickOpenService: IQuickOpenService,
+		@IActivityService private activityService: IActivityService
+	) {
+		super(id, name);
+	}
+
+	run(): TPromise<any> {
+		const infoForExternalViewlets = this.activityService.getInfoForExternalViewlets();
+
+		const picks: IPickOpenEntry[] = [];
+		for (let viewletId in infoForExternalViewlets) {
+			const { isEnabled, treeLabel } = infoForExternalViewlets[viewletId];
+			picks.push({
+				id: viewletId,
+				label: (isEnabled ? 'Disable ' : 'Enable ') + treeLabel
+			});
+		}
+
+		return TPromise.timeout(50 /* quick open is sensitive to being opened so soon after another */).then(() => {
+			this.quickOpenService.pick(picks, { placeHolder: 'Select Viewlet to toggle', autoFocus: 2 }).then(pick => {
```

Comment:
@octref multiple things:
* nls!
* I suggest to not call it Viewlet but rather similar to the action that it was opened from ("Custom Explorer")
* please use the enum and not the number it represents for `autoFocus`